core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala

```diff
@@ -127,8 +126,11 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
 
       createdContainer match {
         case Some(((actor, data), containerState)) =>
-          busyPool = busyPool + (actor -> data)
-          freePool = freePool - actor
+          //only move to busyPool if max reached
+          if (data.activeActivationCount + 1 >= poolConfig.maxConcurrent) {
```

Comment:
Don't know where to put this, so doing this here:

I have one general question on the state-keeping: I think we update the state in the ContainerProxy asynchronously. That is, the proxy sends a "WarmedData" message, which we then use to update state.

Do we need to "predict" that update in the ContainerPool, as in: increment WarmedData on scheduling locally and updating that with the "actual" value once the event happens from ContainerProxy?

I'm mostly wondering if we otherwise have racy counting going on. Thoughts?

Reply:
So the values in ContainerPool will always be stale. If we update the preemptively, they will be accurate momentarily, and inaccurate again after the first in-flight activation completes while the other in-flights are still processing. We could do that, but it seems "even stranger" to not just let the ContainerPool receive the updates. 

Checking for passing the concurrency threshold is "mostly important" for the case where maxConcurrent=1, and in other cases, your mileage will vary as far as "actual concurrency", but pool will try to prevent passing that value, but it will always be working with stale data so will not be 100% precise. I think this is OK. WDYT?